### PR TITLE
`mmt4d` ukernel for the `bf16*bf16->f32` case using AVX-512-BF16

### PR DESCRIPTION
**Pull request by bjacob**
`bf16*bf16->f32` is the one floating-point element type that can perform faster than regular `f32` on widely-available x86 CPUs, as AVX-512-BF16 is available on Intel Cooper Lake and AMD Zen4 microarchitectures.

By contrast, AVX-512-FP16 is only available on Intel Sapphire Rapids.

The below benchmark results show this being a 2x speedup over `f32` and matching current `i8*i8->i32` performance with VNNI, which could make it quite useful as rounding f32 to the nearest bf16 is easier and less accuracy-compromising than i8-quantizing. The accumulator is still f32 so there's no change there.

### Crazy idea

Rewriting `f32` matmuls into rounding LHS/RHS to `bf16` then doing a `bf16*bf16->f32` matmul is something that we could do as an (opt-in) automatic rewrite for people who have a `f32` workload on their hands and who would like a quick flag to try to get a cheap speedup. Accuracy is likely to be good enough for a majority of workloads, unlike much harder `i8` quantization. Try this flag, get 2x speedup on f32 workloads on recent x86 CPU, and a 4x speedup on recent ARM. By contrast, an equivalent flag for fp16 would be a  2x slowdown on x86 (except Sapphire Rapids) and a 2x speedup on recent ARM. 

### Benchmark on AMD Zen4 (Ryzen 9 7940HS), `mmt4d_benchmark` (single-thread microbenchmark):

|Benchmark                                              |Gop/s|
|-------------------------------------------------------|-----|
|BM_mmt4d_f32f32f32_tile_1x8x1_avx2_fma                 |25.5 |
|BM_mmt4d_f32f32f32_tile_2x8x1_avx2_fma                 |43.3 |
|BM_mmt4d_f32f32f32_tile_4x8x1_avx2_fma                 |82.3 |
|BM_mmt4d_f32f32f32_tile_8x8x1_avx2_fma                 |132.9|
|BM_mmt4d_f32f32f32_tile_1x16x1_avx512_base             |44.1 |
|BM_mmt4d_f32f32f32_tile_2x16x1_avx512_base             |87.8 |
|BM_mmt4d_f32f32f32_tile_4x16x1_avx512_base             |167.1|
|BM_mmt4d_f32f32f32_tile_8x16x1_avx512_base             |166.7|
|BM_mmt4d_f32f32f32_tile_16x16x1_avx512_base            |166.3|
|BM_mmt4d_f16f16f32_tile_1x8x1_avx2_fma                 |21.6 |
|BM_mmt4d_f16f16f32_tile_2x8x1_avx2_fma                 |41.0 |
|BM_mmt4d_f16f16f32_tile_4x8x1_avx2_fma                 |74.1 |
|BM_mmt4d_f16f16f32_tile_8x8x1_avx2_fma                 |81.0 |
|BM_mmt4d_f16f16f32_tile_1x16x1_avx512_base             |38.6 |
|BM_mmt4d_f16f16f32_tile_2x16x1_avx512_base             |64.5 |
|BM_mmt4d_f16f16f32_tile_4x16x1_avx512_base             |73.5 |
|BM_mmt4d_f16f16f32_tile_8x16x1_avx512_base             |79.4 |
|BM_mmt4d_f16f16f32_tile_16x16x1_avx512_base            |82.6 |
|BM_mmt4d_f16f16f16_tile_1x8x1_avx2_fma                 |21.6 |
|BM_mmt4d_f16f16f16_tile_2x8x1_avx2_fma                 |40.9 |
|BM_mmt4d_f16f16f16_tile_4x8x1_avx2_fma                 |73.8 |
|BM_mmt4d_f16f16f16_tile_8x8x1_avx2_fma                 |80.1 |
|BM_mmt4d_f16f16f16_tile_1x16x1_avx512_base             |39.1 |
|BM_mmt4d_f16f16f16_tile_2x16x1_avx512_base             |66.5 |
|BM_mmt4d_f16f16f16_tile_4x16x1_avx512_base             |73.7 |
|BM_mmt4d_f16f16f16_tile_8x16x1_avx512_base             |79.3 |
|BM_mmt4d_f16f16f16_tile_16x16x1_avx512_base            |82.3 |
|BM_mmt4d_bf16bf16f32_tile_1x16x2_avx512_bf16           |68.0 |
|BM_mmt4d_bf16bf16f32_tile_2x16x2_avx512_bf16           |123.2|
|BM_mmt4d_bf16bf16f32_tile_4x16x2_avx512_bf16           |228.9|
|BM_mmt4d_bf16bf16f32_tile_8x16x2_avx512_bf16           |333.6|
|BM_mmt4d_bf16bf16f32_tile_16x16x2_avx512_bf16          |332.2|
|BM_mmt4d_i8i8i32_tile_1x8x2_avx2_fma                   |57.6 |
|BM_mmt4d_i8i8i32_tile_2x8x2_avx2_fma                   |78.3 |
|BM_mmt4d_i8i8i32_tile_4x8x2_avx2_fma                   |92.9 |
|BM_mmt4d_i8i8i32_tile_8x8x2_avx2_fma                   |186.0|
|BM_mmt4d_i8i8i32_tile_1x16x2_avx512_base               |37.8 |
|BM_mmt4d_i8i8i32_tile_2x16x2_avx512_base               |49.0 |
|BM_mmt4d_i8i8i32_tile_4x16x2_avx512_base               |59.2 |
|BM_mmt4d_i8i8i32_tile_8x16x2_avx512_base               |118.4|
|BM_mmt4d_i8i8i32_tile_16x16x2_avx512_base              |231.8|
|BM_mmt4d_i8i8i32_tile_1x16x2_avx512_vnni               |50.8 |
|BM_mmt4d_i8i8i32_tile_2x16x2_avx512_vnni               |70.0 |
|BM_mmt4d_i8i8i32_tile_4x16x2_avx512_vnni               |83.3 |
|BM_mmt4d_i8i8i32_tile_8x16x2_avx512_vnni               |165.5|
|BM_mmt4d_i8i8i32_tile_16x16x2_avx512_vnni              |328.1|

### CODE REVIEW

**Comment by bjacob** [comment]
The CI failures are a LLVM bug causing the Clang bitcode build of this ukernel to infinitely recurse, reported here: https://github.com/llvm/llvm-project/issues/68117

**Comment by bjacob** [comment]
OK, this should be good to do, but while working on this, I encountered and reported 4 issues against LLVM/x86:

* 1 general issue with AVX-512: https://github.com/llvm/llvm-project/issues/68818.
* 3 issues specific to AVX-512-BF16:
  - https://github.com/llvm/llvm-project/issues/68117
  - https://github.com/llvm/llvm-project/issues/68810
  - https://github.com/llvm/llvm-project/issues/68149

**Review by benvanik — APPROVED** [review]

File: runtime/src/iree/builtins/ukernel/arch/x86_64/mmt4d_x86_64_avx512_bf16.c [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  } else {
+    for (int i = 0; i < M0; ++i) {
+      acc[i] = _mm512_setzero_ps();
+    }
+  }
+
+  for (iree_uk_int32_t k = 0; k < params->K; ++k) {
+    __m512bh rhs = bitcast_16xf32_to_32xbf16(_mm512_loadu_ps(rhs_ptr));
+    rhs_ptr += 32;
+#pragma clang loop unroll(full)
```
Comment (jpienaar) [inline runtime/src/iree/builtins/ukernel/arch/x86_64/mmt4d_x86_64_avx512_bf16.c:65]: When compiling with GCC I get

```
/usr/local/google/home/jpienaar/Compiles/iree/runtime/src/iree/builtins/ukernel/arch/x86_64/mmt4d_x86_64_avx512_bf16.c: In function ‘iree_uk_mmt4d_tile_bf16bf16f32_1x16x2_to_16x16x2_x86_64_avx512_bf16’:
/usr/local/google/home/jpienaar/Compiles/iree/runtime/src/iree/builtins/ukernel/arch/x86_64/mmt4d_x86_64_avx512_bf16.c:68: error: ignoring ‘#pragma clang loop’ [-Werror=unknown-pragmas]
   68 | #pragma clang loop unroll(full)
      |
cc1: all warnings being treated as errors
```

I think we've been through this before, but is this something that should be caught in the top level cmake file? E.g., use clang or don't treat warnings as errors? (I mean I don't know how that would work with MSVC)

Comment (bjacob) [reply runtime/src/iree/builtins/ukernel/arch/x86_64/mmt4d_x86_64_avx512_bf16.c:65]: oops, the pragma was not meant to be left in. I'll remove it. Thanks for the heads-up (not clear how this wasn't caught on GCC CI, probably different GCC version).

Comment (bjacob) [reply runtime/src/iree/builtins/ukernel/arch/x86_64/mmt4d_x86_64_avx512_bf16.c:65]: Fix: https://github.com/openxla/iree/pull/15224

**Pull request merged**